### PR TITLE
Add support for ANP and BANP in the explain command

### PR DESCRIPTION
**Pull request by Peac36**
Issue: https://github.com/kubernetes-sigs/network-policy-api/issues/153

Screenshot from 2024-01-14 18-09-48
Screenshot from 2024-01-14 18-10-02

These images are based on the following files:

* https://github.com/kubernetes-sigs/network-policy-api/blob/main/conformance/base/admin_network_policy/core-gress-rules-combined.yaml
* https://github.com/kubernetes-sigs/network-policy-api/blob/main/conformance/base/baseline_admin_network_policy/core-gress-rules-combined.yaml

Also made some modifications to the original requirements:
* Add Ingress/Egress rule names in the action column.
* Add some formatting to the subject column.

### CODE REVIEW

**Comment by astoycos** [comment]
Thanks for this @Peac36! We'll get to it ASAP

File: cmd/policy-assistant/pkg/cli/analyze.go
Comment (mattfenwick) [inline cmd/policy-assistant/pkg/cli/analyze.go:129]: Instead of parsing yaml here, it'd probably be better (in the sense of: simpler, easier to maintain) to create these objects in golang.

File: cmd/policy-assistant/pkg/matcher/explain.go
Comment (mattfenwick) [inline cmd/policy-assistant/pkg/matcher/explain.go:53]: I recommend sorting all output so that it's:

 - deterministic
 - predictable for the reader

**Comment by mattfenwick** [comment]
nice one @Peac36 !  thanks for posting screenshots of the output 👍 👍 !!

I'll defer to @astoycos and @huntergregory <- they're the real approvers , please let me know if you'd like me to take a closer look at anything but overall I feel like this is a good patch!

**Comment by huntergregory** [comment]
Thanks for making this PR @Peac36 ! Have been OOF and then got a concussion, but will aim to review this early next week!

File: cmd/policy-assistant/pkg/kube/labelselector.go
Comment (huntergregory) [inline cmd/policy-assistant/pkg/kube/labelselector.go:110]: Can we keep this and `Match expressions: `? Similarly to my previous comment, leaving this will make the output clearer imo

File: cmd/policy-assistant/pkg/matcher/explain.go
Comment (huntergregory) [inline cmd/policy-assistant/pkg/matcher/explain.go:69]: Can we say "NPv1: All peers allowed" for every NPv1 action?

File: cmd/policy-assistant/pkg/matcher/explain.go
Comment (huntergregory) [inline cmd/policy-assistant/pkg/matcher/explain.go:146]: Since we're doing the same for (B)ANP, may help the reader to be explicit:
```suggestion
		return "NPv1: All peers allowed"
```

File: cmd/policy-assistant/pkg/matcher/explain.go
Comment (huntergregory) [inline cmd/policy-assistant/pkg/matcher/explain.go:150]: nit: could you either add check/panic for BANP or just a clarifying comment that this is BANP in case we add more PolicyKind's later?

**Review by huntergregory — CHANGES_REQUESTED** [review]
Thanks again @Peac36 for tackling this important visualization piece!

You found all the code pieces nicely. The only remaining functionality is handling SameLabels and NotSameLabels.

Also, thanks for looking to improve the formatting. We will need to make sure everything is clear for every case. Along that line, could you add UTs to make sure we have expected outputs for all cases? For NPv1, I'd suggest using the example policies from `cyclonus analyze --use-example-policies` (believe that's the correct arg but you can also check `cyclonus analyze --help`).

**Comment by Peac36** [comment]
Thank you for the input. I'll try to resolve it as soon as possible.

File: cmd/policy-assistant/pkg/kube/labelselector.go
Comment (huntergregory) [reply cmd/policy-assistant/pkg/kube/labelselector.go:110]: Or feel free to change the wording, and we can assess outputs for both the MatchLabels case and MatchExpressions

**Comment by huntergregory** [comment]
Also about the Action column, I like the idea of adding the source there. Taking it a step further, I think we could make it more insightful with an overall verdict. Also, since a policy's rules share the same priority, we could consolidate a policy into a single line like below:
```
Verdict: Allow by ANP
ANP: Allow
    pri=15 (policy1): Allow (ineffective rules: Deny, Skip, Allow, Deny)
    pri=16 (policy3): Deny (ineffective rules: Skip)
    pri=35 (policy2): Skip
BANP: Deny (ineffective rules: Allow, Deny)
```

Another example:
```
Verdict: NPv1 decision or Deny by BANP
ANP: Skip
    pri=15 (policy1): Skip
BANP: Deny
```

Final example:
```
Verdict: NPv1 decision
ANP: Skip
    pri=15 (policy1): Skip
```

I think there often won't be ineffective rules.

### Implementation Ideas
- Sorting ANP by priority
- Create a helper function to get verdict similar to https://github.com/kubernetes-sigs/network-policy-api/blob/9967e867a968a25c7e99cba82783ba77da6f0181/cmd/policy-assistant/pkg/matcher/policy.go#L117

**Comment by Peac36** [comment]
@huntergregory  - I've resolved most of the initial feedback and started working on the new implementation idea but I have a question about Source rules. Should we list all ANPs that affect this subject or we can just drop the column and add the policy name in the action column?

**Comment by huntergregory** [comment]
Thanks @Peac36. I'm thinking to list all ANPs that affect this subject to keep in-line with the previous design. For context, here are those example policies I referenced before, and the previous output for NPv1:
```
$ ./cyclonus analyze --use-example-policies --mode explain
INFO[2024-02-11T16:40:59-08:00] log level set to 'info'
explained policies:
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|  TYPE   |       TARGET       |                            SOURCE RULES                             |           PEER           |       PORT/PROTOCOL       |
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
| Ingress | namespace: default | default/accidental-and                                              | namespace: default       | all ports, all protocols  |
|         | Match labels:      | default/accidental-or                                               | pods: Match labels:      |                           |
|         |   a: b             |                                                                     |   role: client           |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: Match labels: |                           |
|         |                    |                                                                     |   user: alice            |                           |
|         |                    |                                                                     | pods: Match labels:      |                           |
|         |                    |                                                                     |   role: client           |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: Match labels: |                           |
|         |                    |                                                                     |   user: alice            |                           |
|         |                    |                                                                     | pods: all                |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-nothing-to-v2-all-web                                 | no pods, no ips          | no ports, no protocols    |
|         | Match labels:      |                                                                     |                          |                           |
|         |   all: web         |                                                                     |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-specific-port-from-role-monitoring-to-app-apiserver   | namespace: default       | port 5000 on protocol TCP |
|         | Match labels:      |                                                                     | pods: Match labels:      |                           |
|         |   app: apiserver   |                                                                     |   role: monitoring       |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-from-app-bookstore-to-app-bookstore-role-api          | namespace: default       | all ports, all protocols  |
|         | Match labels:      |                                                                     | pods: Match labels:      |                           |
|         |   app: bookstore   |                                                                     |   app: bookstore         |                           |
|         |   role: api        |                                                                     |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+                           +
|         | namespace: default | default/allow-from-multiple-to-app-bookstore-role-db                | namespace: default       |                           |
|         | Match labels:      |                                                                     | pods: Match labels:      |                           |
|         |   app: bookstore   |                                                                     |   app: bookstore         |                           |
|         |   role: db         |                                                                     |   role: api              |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: default       |                           |
|         |                    |                                                                     | pods: Match labels:      |                           |
|         |                    |                                                                     |   app: bookstore         |                           |
|         |                    |                                                                     |   role: search           |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: default       |                           |
|         |                    |                                                                     | pods: Match labels:      |                           |
|         |                    |                                                                     |   app: inventory         |                           |
|         |                    |                                                                     |   role: web              |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-nothing                                               | no pods, no ips          | no ports, no protocols    |
|         | Match labels:      |                                                                     |                          |                           |
|         |   app: foo         |                                                                     |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-all-to-app-web                                        | all pods, all ips        | all ports, all protocols  |
|         | Match labels:      | default/allow-all-to-version2-app-web                               |                          |                           |
|         |   app: web         | default/allow-all-to-version3-app-web                               |                          |                           |
|         |                    | default/allow-all-to-version4-app-web                               |                          |                           |
|         |                    | default/allow-from-anywhere-to-app-web                              |                          |                           |
|         |                    | default/allow-from-namespace-to-app-web                             |                          |                           |
|         |                    | default/allow-from-namespace-with-labels-type-monitoring-to-app-web |                          |                           |
|         |                    | default/allow-nothing-to-app-web                                    |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+                           +
|         | namespace: default | default/allow-all-within-namespace                                  | namespace: default       |                           |
|         | all pods           | default/allow-nothing-to-anything                                   | pods: all                |                           |
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         |                    |                                                                     |                          |                           |
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
| Egress  | namespace: default | default/allow-egress-on-port-app-foo                                | all pods, all ips        | port 53 on protocol TCP   |
|         | Match labels:      | default/allow-egress-to-all-namespace-from-app-foo-on-port-53       |                          | port 53 on protocol UDP   |
|         |   app: foo         | default/allow-no-egress-from-labels-app-foo                         |                          |                           |
|         |                    | default/allow-nothing                                               |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-no-egress-from-namespace                              | no pods, no ips          | no ports, no protocols    |
|         | all pods           |                                                                     |                          |                           |
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
```

**Comment by huntergregory** [comment]
@Peac36 btw would you mind continuing the PR in separate commits? Will make it easier to review/track changes

**Comment by huntergregory** [comment]
Also @Peac36, we're planning to record a demo in a week or so for the Policy Assistant. Would be awesome to showcase your work! I'd like to give you a shoutout in a slide for our upcoming KubeCon presentation, similar to [slide 25 here](https://static.sched.com/hosted_files/kccncna2023/af/Network%20Policy%20API_%20Intro%20and%20Project%20Update.pptx.pdf) from the Fall.

Feel free to join the Kubernetes Slack workspace -> [sig-network-policy-api channel](https://kubernetes.slack.com/archives/C01CWSHQWQJ) and DM me there!

**Comment by Peac36** [comment]
>@Peac36 btw would you mind continuing the PR in separate commits? Will make it easier to review/track changes

Sure.

Also @Peac36, we're planning to record a demo in a week or so for the Policy Assistant. Would be awesome to showcase your work! I'd like to give you a shoutout in a slide for our upcoming KubeCon presentation, similar to [slide 25 here](https://static.sched.com/hosted_files/kccncna2023/af/Network%20Policy%20API_%20Intro%20and%20Project%20Update.pptx.pdf) from the Fall.

Awesome. I'm almost ready with the grouping of the ANPs and after that, I'll take care of the BANPs and the tests. I will be ready with the whole thing probably on Monday. I hope this is OK for you guys.

**Comment by Peac36** [comment]
Hey @huntergregory I'm ready with the feedback and implementation of the idea you gave.

Here is a short preview.

Screenshot from 2024-02-26 20-29-13

Added tests for `BuildV1AndV2NetPols` to assert that

* Anps with the same target would be grouped together
* Anps that have the same target,protocols, and peers would be together
* Banps would be added to a target group if it has the same peers and protocols.

Let me know if you have more ideas for tests.

File: cmd/policy-assistant/pkg/matcher/builder.go
Comment (huntergregory) [reply cmd/policy-assistant/pkg/matcher/builder.go:42]: https://github.com/kubernetes-sigs/network-policy-api/blob/dc00b3607953b5a11062ac2e25f1a9e1e2893164/cmd/policy-assistant/pkg/matcher/simplifier.go#L19

Comment (huntergregory) [reply cmd/policy-assistant/pkg/matcher/builder.go:42]: if all other comments make sense, I think we can just revert all changes to this file?

File: cmd/policy-assistant/pkg/matcher/builder_tests.go
Comment (huntergregory) [inline cmd/policy-assistant/pkg/matcher/builder_tests.go:352]: can keep this test case and set simplify=true

File: cmd/policy-assistant/pkg/matcher/explain.go
Comment (huntergregory) [reply cmd/policy-assistant/pkg/matcher/explain.go:85]: Within this function call, we can group by PeerMatchers by Peer and Port

**Review by huntergregory — CHANGES_REQUESTED** [review]
Thanks @Peac36 the screenshot looks really good. Can we flip the ordering for priorities in Action column? Highest priority (lowest pri number) coming first.

Had a few thoughts on simplifying implementation. My main goal is to keep parallels with NPv1 where possible, and to minimize change to NPv1 and functionality outside the summary table.

Also, thanks for adding tests. So I was thinking to prevent regressions, it'd be nice to just have maybe one integration test comparing the output of `cyclonus analyze --mode explain` to an expected table (including NPv1).

Pretty simple, just something like this as a regular UT (no need for gomega?) in a new file _test/integration/explain_test.go_:
```
tableOutput := table.Render()
expectedTable := """---------------------------
| Name | Subject | Action |
| abc     | ns=x     | allow   |
---------------------------
require.Equal(t, expectedTable, tableOutput)
"""
```

Also, we don't have a PR check for it yet. Could you run `go test .` in both _test/integration/_ and _pkg/matcher/_ directories to make sure previous tests are all green?

**Review by huntergregory — COMMENTED** [review]
Thanks for simplifying @Peac36, this will help us in the long run. And thanks for working on this quickly. Let's address a couple questions/nits and then we can get this merged. *Would you be able to tackle these ASAP?*

Seeing a few discrepancies between your last screenshot and the text in your integration test. Do you know why? Which is correct?

Before:
image

After:
image

**Comment by Peac36** [comment]
Thanks for the feedback @huntergregory. Already fixed some of them, however, I'll need some more - probably later tomorrow will be able to resolve them all. About the discrepancies - the issue came from here:

https://github.com/kubernetes-sigs/network-policy-api/pull/188/files#diff-0c40ee241d83fc2c8e730d1cfea6a20eccbfd5a0c99a346073bee43e6646c44dR194

I missed including the namespace selector when building the map key so the grouping was incorrect.

Here is the result after the fix. Please note that I made some changed to the Banp so the result should not be the same as the one from previous examples. 
Screenshot from 2024-03-06 18-42-13

**Comment by Peac36** [comment]
@huntergregory - I've resolved your last feedback.

**Review by huntergregory — APPROVED** [review]
Awesome thank you @Peac36 !!

Lgtm. @astoycos could we get your seal of approval?

**Comment by huntergregory** [comment]
Andrew added me as owner actually, let’s see if this works 😆 

/lgtm

/approve

**Comment by mattfenwick** [comment]
> Awesome thank you @Peac36 !!
> 
> /lgtm
> 
> @mattfenwick could we get your seal of approval? Also going to raise a PR to add myself to [cmd/policy-assistant/OWNERS](https://github.com/kubernetes-sigs/network-policy-api/blob/main/cmd/policy-assistant/OWNERS).

I know it's too late but I'll do it anyway!

/lgtm
/approve

great work @Peac36 , I love where y'all are taking this!

**Comment by huntergregory** [comment]
Congrats @Peac36 on the first policy assistant PR!

Just messaged you on Slack about something for tomorrow (Monday) if you have time

**Comment by huntergregory** [comment]
Fixed #153

**Pull request merged**
